### PR TITLE
[hotfix][runtime] Add job ID to RuntimeContext

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

There are some [cases](http://apache-flink-user-mailing-list-archive.2336050.n4.nabble.com/How-to-get-flink-JobId-in-runtime-td36756.html) when job ID needs to be accessed from the user code (function).
Existing workarounds doesn't look clean (reliable).

This PR adds `Optional<JobID>` to the `RuntimeContext` (the latter already contains some information of the same level, such as subtask index).

cc: @aljoscha, @StephanEwen 

## Verifying this change

This change is a trivial rework without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **yes**
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java [hunk 1]
```diff
@@ -56,6 +58,13 @@
 @Public
 public interface RuntimeContext {
 
+    /**
+     * The ID of the current job. Empty if the execution happens outside of any job context (e.g.
+     * standalone collection executor). The returned ID should NOT be used for any job management
```
Comment (AHeise) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java:63]: I'd add that JobID may change upon (manual?) restart.

**Review by AHeise — COMMENTED** [review]
Thank you very much for your contribution. I'm wondering if we can add simple ITCase to verify that the job id is actually correctly set.

File: flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java (hunk 1 above)
Comment (rkhachatryan) [reply flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java]: Done.

**Comment by rkhachatryan** [comment]
AZP failure not related (FLINK-17510)

**Review by AHeise — APPROVED** [review]
LGTM

**Comment by rkhachatryan** [comment]
Thanks for reviewing @AHeise.
Merging.

**Comment by kezhuw** [comment]
Hi @rkhachatryan, I am a bit worry about whether the type `Optional<JobID>` is a right choice for reasons:

* Most `getJobId` expect a not nullable `JobID`.
* The only path `CollectionEnvironment`, `CollectionExecutor` could also be shaped to return a plan global job id(or created before execution if `Plan.jobId` is null). Due to `getJobId` is a brand new api, it does not hurt anyone.
* After `DataSet` phased out, that only path will also be phased out. Then, it will be really confused that `getJobId` returns an optional value in that stage.

cc @AHeise  @aljoscha  @StephanEwen

**Comment by StephanEwen** [comment]
+1 to the comment from @kezhuw , I also marked the corresponding JIRA ticket a blocker.

**Pull request merged**